fondant/dataset.py

```diff
@@ -95,71 +114,145 @@ def load_dataframe(self, spec: FondantComponentSpec) -> dd.DataFrame:
+            Verify that all the fields defined in the output subset are present in
+            the input dataframe
+            """
 
-    def add_subsets(self, df: dd.DataFrame, spec: FondantComponentSpec):
-        for name, subset in spec.output_subsets.items():
-            fields = list(subset.fields.keys())
-            # verify fields are present in the output dataframe
-            subset_columns = [f"{name}_{field}" for field in fields]
+            field_names = list(subset_fields.keys())
+            subset_columns = [f"{subset_name}_{field}" for field in field_names]
             subset_columns.extend(self.mandatory_subset_columns)
 
             for col in subset_columns:
                 if col not in df.columns:
                     raise ValueError(
-                        f"Field {col} defined in output subset {name} but not found in dataframe"
+                        f"Field {col} defined in output subset {subset_name} "
+                        f"but not found in dataframe"
                     )
 
-            # load subset dataframe
+            return subset_columns
+
+        def create_subset_dataframe(subset_name, subset_columns, df):
+            """
+            Create subset dataframe to save with the original field name as the column name
+            """
+            # Create a new dataframe with only the columns needed for the output subset
             subset_df = df[subset_columns]
-            # remove subset prefix from subset columns
+
+            # Remove the subset prefix from the column names
+            prefix_to_replace = f"{subset_name}_"
             subset_df = subset_df.rename(
                 columns={
-                    col: col.split("_")[-1]
+                    col: col.replace(prefix_to_replace, "")
                     for col in subset_df.columns
                     if col not in self.mandatory_subset_columns
+                    and col.startswith(prefix_to_replace)
                 }
             )
-            # add to the manifest and upload
-            self._upload_subset(name, subset.fields, subset_df)
 
-    def upload(self, save_path):
+            return subset_df
+
+        upload_subsets_task = []
+
+        # Loop through each output subset defined in the spec
+        for subset_name, subset in spec.output_subsets.items():
+            # Verify that all the fields defined in the output subset are present in the
+            # input dataframe
+            subset_columns = verify_subset_columns(subset_name, subset.fields, df)
+
+            # Create a new dataframe with only the columns needed for the output subset
+            subset_df = create_subset_dataframe(subset_name, subset_columns, df)
+
+            # Add the output subset to the manifest
+            manifest_fields = [
+                (field.name, Type[field.type.name]) for field in subset.fields.values()
+            ]
+            self.manifest.add_subset(subset_name, fields=manifest_fields)
```

Comment:
The `Dataset` class should be a wrapper around an immutable `Manifest` (see discussion [here](https://github.com/ml6team/fondant/pull/56#discussion_r1180159365)).

The `evolve` method currently takes care of updating the manifest (adding subsets based on the output subsets of the component spec). cc @RobbeSneyders